### DISCUSSION TITLE
how to use PyModule multiply times in different parts of code

### DISCUSSION THREAD

**Discussion by g992** [PyO3/pyo3#3824] [Questions] [answered]
I am creating a Py Module from the code (PyModule::from_code) and it takes a very long time, while the module does not change in any way. Can I somehow create a module and return it for further use in the code? I use py.run, where only user_code and locals change. I'm new in rust, so please understand :)

**Comment by davidhewitt** [comment]
Without seeing the specifics of your program, I suggest you can create the module once and pass it around your program as a `Py<PyModule>`, to re-use that one module object.

**Reply by g992** [reply]
I have code like this: 
```rust
fn get_python()-> (PyResult<&'static PyModule>, &'static Python){
    let indicators_file = fs::read_to_string(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/python_lib/base.py"
    ))
        .expect("Should have been able to read the file");

    let (module, &py) = Python::with_gil(|py| {
        let module = PyModule::from_code(py, &indicators_file, "indicators", "indicators");
        (module, &py)
    });

    (module, &py)
}
```
And i want create python with this module once, and use it in my program many times. But! i cant do this because py's lifetime is only in brackets.
I can`t create module every time, because it use around 200ms,

**Reply by davidhewitt** [reply]
Try something more like this:

```rust
fn get_python() -> PyResult<Py<PyModule>> {
    let indicators_file = fs::read_to_string(concat!(
    env!("CARGO_MANIFEST_DIR"),
    "/python_lib/base.py"
    ))
        .expect("Should have been able to read the file");

    Python::with_gil(|py| {
        PyModule::from_code(py, &indicators_file, "indicators", "indicators").into()
    })
}
```

To create the `Py<PyModule>` as I suggest, and then you can pass that around your program.

**Reply by g992** [reply]
It works! Thanks!

**Reply by g992** [reply]
UPD: Its not works..
```
error[E0277]: the trait bound `Result<pyo3::Py<pyo3::prelude::PyModule>, PyErr>: From<Result<&pyo3::prelude::PyModule, PyErr>>` is not satisfied
  --> src/main.rs:60:79
   |
60 |         PyModule::from_code(py, &indicators_file, "indicators", "indicators").into()
   |                                                                               ^^^^ the trait `From<Result<&pyo3::prelude::PyModule, PyErr>>` is not implemented for `Result<pyo3::Py<pyo3::prelude::PyModule>, PyErr>`
   |
   = note: required for `Result<&pyo3::prelude::PyModule, PyErr>` to implement `Into<Result<pyo3::Py<pyo3::prelude::PyModule>, PyErr>>`
```
I try to fix by myself, but zero effect

**Reply by adamreichold** [reply, marked as answer]
Try mapping the value only, e.g. `PyModule::from_code(py, &indicators_file, "indicators", "indicators").map(|module| module.into())`

**Reply by g992** [reply]
Well... mapping helps, but how i can transform module to PyDict? just .dict() not works
Or how can i use it with py.run(...)? I Think conversion to dict not better idea

**Reply by g992** [reply]
I write next code:
```rust
        let globals = PyDict::new(py);
        globals.set_item("f", module)?;
```
this is ok?
